ui/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/ManagementActor.scala

```diff
@@ -120,11 +120,33 @@ class ManagementActor(managers: Map[ProcessingType, ProcessManager],
   //TODO: In future we should move this functionality to ProcessManager.
   private def handleObsoleteStatus(processStateDefinitionManager: ProcessStateDefinitionManager, processState: Option[ProcessState], lastAction: Option[ProcessAction]): ProcessStatus =
     (processState, lastAction) match {
-      case (Some(state), _)  => ProcessStatus.create(state, lastAction)
+      case (Some(state), _)  => handleMismatchStateVersion(processStateDefinitionManager, state, lastAction)
       case (None, Some(action)) if action.isCanceled => ProcessStatus(SimpleStateStatus.Canceled, processStateDefinitionManager)
+      case (None, Some(act)) if act.isDeployed => ProcessStatus.error(
+        SimpleStateStatus.ErrorShouldRunning, processStateDefinitionManager, act.user, act.processVersionId
+      )
       case (None, None) => ProcessStatus(SimpleStateStatus.NotDeployed, processStateDefinitionManager)
     }
 
+  //This method handles some corner cases for mismatch states version
+  //TODO: In future we should move this functionality to ProcessManager.
+  private def handleMismatchStateVersion(processStateDefinitionManager: ProcessStateDefinitionManager, processState: ProcessState, lastAction: Option[ProcessAction]): ProcessStatus = {
+    (processState.version, lastAction) match {
+      case (Some(sv), Some(act)) if act.isDeployed && sv.versionId != act.processVersionId => ProcessStatus.error(
+        SimpleStateStatus.ErrorMismatchVersion, processStateDefinitionManager, act.user, act.processVersionId, Option(processState), Option(sv.versionId)
+      )
+      case (Some(_), Some(act)) if act.isDeployed && !processState.status.isFollowingDeployAction => ProcessStatus.error(
+        SimpleStateStatus.ErrorShouldRunning, processStateDefinitionManager, act.user, act.processVersionId, Option(processState)
+      )
+      case (Some(stateVersion), Some(action)) if stateVersion.versionId == action.processVersionId =>
+        ProcessStatus(processState)
+      case (Some(sv), None) if processState.isDeployed => ProcessStatus.error(
+        SimpleStateStatus.ErrorShouldNotBeDeployed, processStateDefinitionManager, sv.user, sv.versionId, Option(processState)
+      )
+      case _ => ProcessStatus(processState)
```

Comment:
What cases are here? Can you remove this line or compiler return "match is not exhaustive"? If so, maybe we can explicitly write them down and think a bit if they are really normal, correct cases. It they aren't we should return other state or throw exception. IMO e.g. `(None, Some(_))` isn't normal.